src/ir/ReFinalize.cpp

```diff
@@ -139,9 +141,107 @@ void ReFinalize::visitRefTest(RefTest* curr) { curr->finalize(); }
 void ReFinalize::visitRefCast(RefCast* curr) { curr->finalize(); }
 void ReFinalize::visitBrOn(BrOn* curr) {
   curr->finalize();
-  if (curr->type == Type::unreachable) {
-    replaceUntaken(curr->ref, nullptr);
-  } else {
+  if ((curr->op == BrOnCast || curr->op == BrOnCastFail) &&
+      !Type::isSubType(curr->castType, curr->ref->type)) {
+    // If the input type has been refined such that it is no longer a supertype
+    // of the cast type, then this can no longer be a valid br_on_cast*. Replace
+    // it with someting else. Even though these optimizations could apply to
+    // valid br_on_*, only apply them to invalid br_on_* here to avoid
+    // validation failures after parsing valid but optimizable modules. These
+    // optimizations are applied more widely in RemoveUnusedBrs.
+    auto brCond =
+      GCTypeUtils::evaluateCastCheck(curr->ref->type, curr->castType);
+    if (curr->op == BrOnCastFail) {
+      brCond = GCTypeUtils::flipEvaluationResult(brCond);
+    }
+    Builder builder(*getModule());
+    switch (brCond) {
+      case GCTypeUtils::Unknown: {
+        // The branch may or may not be taken, so the cast must be valid.
```

Comment:
We've checked the subtyping above such that we only get down to this switch if the input type is not a supertype of the cast type, but that's the only situation in which the result of the cast can be completely unknown until runtime. In other words, this is the good case that doesn't need fixing.

We still need the separate check above, though, because while the cast result being unknown implies that the input is a supertype of the output, the reverse implication does not hold.